File: 414.third-maximum-number.py
#
# @lc app=leetcode id=414 lang=python3
#
# [414] Third Maximum Number
#

# @lc code=start
class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        first_max = -float('inf')
        second_max = -float('inf')
        third_max = -float('inf')

        for val in nums:
            if val == first_max or val == second_max or val == third_max:
                continue

            if val > first_max:
                third_max = second_max
                second_max = first_max
                first_max = val

            elif val > second_max:
                third_max = second_max
                second_max = val

            elif val > third_max:
                third_max = val

        
        if third_max != -float('inf'):
            return third_max
        else:
            return first_max

        
# @lc code=end

File: 414.third-maximum-number.java
/*
 * @lc app=leetcode id=414 lang=java
 *
 * [414] Third Maximum Number
 */

// @lc code=start
class Solution {
    public int thirdMax(int[] nums) {
        Integer firstMax = null;
        Integer secondMax = null;
        Integer thirdMax = null;

        for (Integer val : nums) {
            if (val.equals(firstMax) || val.equals(secondMax) || val.equals(thirdMax)) {
                continue;
            }


            if (firstMax == null || val > firstMax) {
                thirdMax = secondMax;
                secondMax = firstMax;
                firstMax = val;
            } else if (secondMax == null || val > secondMax) {
                thirdMax = secondMax;
                secondMax = val;
            } else if (thirdMax == null || val > thirdMax) {
                thirdMax = val;
            }
        }



        if (thirdMax != null) {
            return thirdMax;
        } else {
            return firstMax;
        }
    }
}
// @lc code=end

